designs/bare-metal-out-of-band-webhook.md

```diff
@@ -0,0 +1,549 @@
+# Python example
+import datetime
+import hmac
+import hashlib
+import json
+
+# 1. concatenate the body and the timestamp header
+body = json.loads('{"host":"192.168.2.3","task":{"bootDevice":{"device":"pxe","persistent":false,"efiBoot":false}}}')
+headers = {"X-Rufio-Timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat(), "Content-Type": "application/json"}
+signPayload = '{}{}'.format(json.dumps(req.json), req.headers.get('X-Rufio-Timestamp')).encode('utf-8')
+
+# 2. HMAC sign
+secret = bytes('superSecret1' , 'utf-8')
+signature = hmac.new(secret, signPayload, hashlib.sha256)
+
+# 3. Make a hex encoded string of the HMAC signature
+sig = signature.hexdigest()
+
+# 4. Prepend the algorithm type to the signature
+sig = 'sha256=' + sig
+
+# 5. Store the string signature in the header
+headers["X-Rufio-Signature-256"] = sig
+```
+
+#### Full HTTP Request Example
+
+Will we always use `application/json` as the content type. HTTP method will always be `POST`
+
+```text
+POST /webhook
+Host: www.example.com
+Content-Type: application/json
+Timestamp: 2023-06-09T10:26:10-06:00
+X-Rufio-Signature: 2dce505d96a53c5768052ee90f3df2055657518dad489160df9913f66042e160
+{
+    "host": "192.168.2.2",
+    "task": {
+        "bootDevice": {
+            "device": "pxe",
+            "persistent": false,
+            "efiBoot": false
+        }
+    }
+}
+```
+
+#### EKS Anywhere integration
+
+Up to this point in the doc all the development has been in the Rufio code base. Once all that code is complete and merged, in EKS Anywhere we will update the `[GIT_TAG](https://github.com/aws/eks-anywhere-build-tooling/blob/main/projects/tinkerbell/rufio/GIT_TAG)`  in the [build-tooling](https://github.com/aws/eks-anywhere-build-tooling/) repo for Rufio. We will also need to handle the new environment variables that need to be passed to Rufio in the [Tinkerbell Helm chart](https://github.com/aws/eks-anywhere-build-tooling/tree/main/projects/tinkerbell/tinkerbell-chart).
+
+* `webhook_provider_secrets`
+* `webhook_consumer_url`
+* `webhook_consumer_tls_cert`
+
+Currently BMC credentials are stored in a csv file and are per machine. The paradigm for the webhook provider is a bit different. There will not be a user/password combo per machine, instead all machines will use the secret(s) provided in `webhook_provider_secrets`.
+
+The webhook provider will require at a minimum 2 things from the customer.
+
+1. One or more HMAC secrets for use in signing the payload (HMAC).
+    1. These secrets must be provided as a comma separated environment variable. `webhook_provider_secrets="secretOne,secretTwo,secretThree,etc"`
+    2. A minimum complexity and length for each HMAC secret will be required and validated.
+2. A URL for where we will post webhook notifications.
+    1. This URL must be provided as an environment variable. `webhook_consumer_url=https://mywebhookconsumer/webhooks"`
+    2. We will do a preflight validation check for connectivity to this URL.
```

Comment:
How do we envisage this working? Would we need to require webhook providers to implement a readiness type endpoint possibly configured via a `webhook_consumer_readiness_url` environment variable.

Reply:
Great point here, thanks. In order to satisfy the bmclib interfaces we will implement the `Open` method. This is where i envision we make a call to the `webhook_consumer_url` to see if it's reachable. I understand this is a runtime call mid BMC interaction and maybe in an ideal world this `webhook_consumer_url` would be checked as early as possible but this is not the model that bmclib has. Im open to discuss alternatives/options/etc.